src/test/rsm_tests.cpp

```diff
@@ -0,0 +1,201 @@
+
+#include "recursive_shared_mutex.h"
+#include "utiltime.h"
+
+#include "test/test_bitcoin.h"
+
+#include <boost/test/unit_test.hpp>
+
+
+BOOST_FIXTURE_TEST_SUITE(rsm_tests, BasicTestingSetup)
+
+recursive_shared_mutex rsm;
+std::vector<int> rsm_guarded_vector;
+
+void alpha()
+{
+    // lock and add a number
+    rsm.lock(std::this_thread::get_id());
+    rsm_guarded_vector.push_back(0);
+
+    // recursive lock and add another number
+    rsm.lock(std::this_thread::get_id());
+    rsm_guarded_vector.push_back(1);
+
+    // lock shared and add a number while holding write lock
+    rsm.lock_shared(std::this_thread::get_id());
+    rsm_guarded_vector.push_back(2);
```

Comment:
Not sure if you covered it , but I didn't see that you did two or more shared locks in a row...the locks I see but just not the shared_locks.  Would also be good I think to do more than just 2 in a row...could do many in a loop...(what are the limits , are there any?)

Reply:
added test that recursively locks 100 times to address this concern. i recall reading somewhere that there is an unknown internal limit in c++  for mutex locking but it is higher than should ever be hit by proper normal use.